### PR TITLE
feat(Arbitrum_Adapter): Support custom gas token

### PR DESCRIPTION
**Pull request by nicholaspai**
- Example `createRetryableTicket` paying for submission with custom gas token and also sending custom gas token to L2: https://etherscan.io/tx/0xfbcd1f29a0db9971c885815e74499f3a545e7c0ee0a2d06ee4a94dc08c03f4e0

- Example `outboundTransfer` paying with custom gas token and sending a different ERC20 to L2: https://etherscan.io/tx/0x7f65fff856520cc419c3a9f43a61f3427ad495ddf5dc81276c2767ac74242e39

### CODE REVIEW

File: contracts/chain-adapters/Arbitrum_Adapter.sol [hunk 1]
```diff
@@ -233,54 +321,93 @@ contract Arbitrum_Adapter is AdapterInterface, CircleCCTPAdapter {
...
-                    l1Token,
-                    to,
-                    amount,
-                    RELAY_TOKENS_L2_GAS_LIMIT,
-                    L2_GAS_PRICE,
-                    data
-                );
+
+            // If custom gas token, call special functions that handle paying with custom gas tokens.
+            if (address(CUSTOM_GAS_TOKEN) != address(0)) {
```
Comment (mrice32) [inline contracts/chain-adapters/Arbitrum_Adapter.sol:327]: Instead of these huge if else blocks, do you think it's worthwhile to just make a different adapter for these rather than squeeze them into the existing arb adapter?

File: contracts/chain-adapters/Arbitrum_Adapter.sol [hunk 2]
```diff
@@ -168,6 +225,13 @@ contract Arbitrum_Adapter is AdapterInterface, CircleCCTPAdapter {
 
     ArbitrumL1ERC20GatewayLike public immutable L1_ERC20_GATEWAY_ROUTER;
 
+    // This token can be used to pay for l1 to l2 messages if its configured by an Arbitrum orbit chain.
+    // If this is initialized to be non-zero, then this token will be pulled from the customL2GasTokenFunder contract
+    // to pay for l1 to l2 messages.
+    IERC20 public immutable CUSTOM_GAS_TOKEN;
+
+    address public immutable CUSTOM_GAS_TOKEN_FUNDER;
```
Comment (mrice32) [inline contracts/chain-adapters/Arbitrum_Adapter.sol:233]: Will this contract be added in a separate PR?

**Review by mrice32 — APPROVED** [review]
Looks great! Just two questions

File: contracts/chain-adapters/Arbitrum_Adapter.sol (hunk 1 above)
Comment (nicholaspai) [reply contracts/chain-adapters/Arbitrum_Adapter.sol:327]: hmm there is a lot of code we can re-use though

File: contracts/chain-adapters/Arbitrum_Adapter.sol (hunk 2 above)
Comment (nicholaspai) [reply contracts/chain-adapters/Arbitrum_Adapter.sol:233]: yeah let's just assume for now it is some contract that has granted approval to this contract to pull all of its funds. WDYT?

Comment (mrice32) [reply contracts/chain-adapters/Arbitrum_Adapter.sol:233]: Do you think we need to audit that contract?

File: contracts/chain-adapters/Arbitrum_Adapter.sol (hunk 1 above)
Comment (mrice32) [reply contracts/chain-adapters/Arbitrum_Adapter.sol:327]: So I guess there are three options:
1. Put them all in the same adapter, as we have here.
2. Copy the code that is identical and make a separate custom gas adapter.
3. Create a common base class to share the common code.

Comment (mrice32) [reply contracts/chain-adapters/Arbitrum_Adapter.sol:327]: I think 2 is the approach that makes the adapters the easiest to reason about, but not super strongly opinionated.

**Review by mrice32 — COMMENTED** [review]
Left follow-up comments

File: contracts/chain-adapters/Arbitrum_Adapter.sol (hunk 1 above)
Comment (nicholaspai) [reply contracts/chain-adapters/Arbitrum_Adapter.sol:327]: 2 is also slightly cheaper gas cost wise, i'm ok with that. Makes the diff easier to reason about

File: contracts/chain-adapters/Arbitrum_Adapter.sol (hunk 2 above)
Comment (nicholaspai) [reply contracts/chain-adapters/Arbitrum_Adapter.sol:233]: probably but i didn't want to build it in this PR. It could even be an EOA that we seed with funds, right?

Comment (nicholaspai) [reply contracts/chain-adapters/Arbitrum_Adapter.sol:233]: i think the key thing to center around is whether we should "pull" from the funder via transferFrom or should there be an interface that exposes a function like `pull()` that we can cull to get funds

File: contracts/chain-adapters/Arbitrum_Adapter.sol (hunk 1 above)
Comment (nicholaspai) [reply contracts/chain-adapters/Arbitrum_Adapter.sol]: https://github.com/across-protocol/contracts/pull/585/commits/d6f038d8bf878065cfb9af2821f5428af77cca20

File: contracts/chain-adapters/Arbitrum_Adapter.sol (hunk 2 above)
Comment (nicholaspai) [reply contracts/chain-adapters/Arbitrum_Adapter.sol]: I guess one problem with just calling transferFrom from a set address is that LP's might worry that the address gets set to their EOA, which has approved the HubPool to spend their tokens, and can be used as a vector to take their funds.

Comment (nicholaspai) [reply contracts/chain-adapters/Arbitrum_Adapter.sol]: wdyt of this simple funder: edc3fd86b226a7b6d5321f9853c91f28d5a9c8f7

Comment (mrice32) [reply contracts/chain-adapters/Arbitrum_Adapter.sol]: Yeah, I like that we're not just doing a transferFrom from an arbitrary address. +1 for simple funder contract.

File: contracts/chain-adapters/Arbitrum_CustomGasToken_Adapter.sol [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+    function getL1CallValue(uint32 l2GasLimit) public view returns (uint256) {
+        return L2_MAX_SUBMISSION_COST + L2_GAS_PRICE * l2GasLimit;
+    }
+
+    function _pullCustomGas(uint32 l2GasLimit) internal returns (uint256) {
+        uint256 requiredL1CallValue = getL1CallValue(l2GasLimit);
+        CUSTOM_GAS_TOKEN_FUNDER.withdraw(CUSTOM_GAS_TOKEN, requiredL1CallValue);
+        require(CUSTOM_GAS_TOKEN.balanceOf(address(this)) >= requiredL1CallValue, "Insufficient gas balance");
+        return requiredL1CallValue;
+    }
```
Comment (mrice32) [inline contracts/chain-adapters/Arbitrum_CustomGasToken_Adapter.sol:290]: Just to be really explicit: the reason we don't want to just pull these tokens from msg.sender is that these executions are often called via the external multicall contract, so this would mean that we could no longer batch these executions with the mainnet spoke pool relayer refund leaf execution, right?

File: contracts/chain-adapters/Arbitrum_CustomGasToken_Funder.sol [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    using SafeERC20 for IERC20;
+
+    /**
+     * @notice Deposit tokens into the contract.
+     * @param token Token to deposit.
+     * @param amount Amount of tokens to deposit.
+     */
+    function deposit(IERC20 token, uint256 amount) external {
+        token.safeTransferFrom(msg.sender, address(this), amount);
+    }
```
Comment (mrice32) [inline contracts/chain-adapters/Arbitrum_CustomGasToken_Funder.sol:18]: nit: do we need a deposit function if we can just transfer tokens in?

File: contracts/chain-adapters/Arbitrum_CustomGasToken_Adapter.sol [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        FunderInterface _customGasTokenFunder,
+        uint256 _l2MaxSubmissionCost
+    ) CircleCCTPAdapter(_l1Usdc, _cctpTokenMessenger, CircleDomainIds.Arbitrum) {
+        L1_INBOX = _l1ArbitrumInbox;
+        L1_ERC20_GATEWAY_ROUTER = _l1ERC20GatewayRouter;
+        L2_REFUND_L2_ADDRESS = _l2RefundL2Address;
+        CUSTOM_GAS_TOKEN = _customGasToken;
+        L2_MAX_SUBMISSION_COST = _l2MaxSubmissionCost;
+        CUSTOM_GAS_TOKEN_FUNDER = _customGasTokenFunder;
+        require(L1_INBOX.bridge().nativeToken() == address(_customGasToken), "Incorrect custom gas token");
```
Comment (mrice32) [inline contracts/chain-adapters/Arbitrum_CustomGasToken_Adapter.sol:187]: nit: why pass _customGasToken and check it vs just reading it from nativeToken()?

File: contracts/chain-adapters/Arbitrum_CustomGasToken_Adapter.sol [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+                    L2_MAX_SUBMISSION_COST, // maxSubmissionCost Max gas deducted from user's L2 balance to cover base fee
+                    L2_REFUND_L2_ADDRESS, // excessFeeRefundAddress maxgas * gasprice - execution cost gets credited here on L2
+                    L2_REFUND_L2_ADDRESS, // callValueRefundAddress l2Callvalue gets credited here on L2 if retryable txn times out or gets cancelled
+                    RELAY_MESSAGE_L2_GAS_LIMIT, // maxGas Max gas deducted from user's L2 balance to cover L2 execution
+                    L2_GAS_PRICE, // gasPriceBid price bid for L2 execution
+                    amountToBridge, // tokenTotalFeeAmount amount of fees to be deposited in native token.
+                    "0x" // data ABI encoded data of L2 message
+                );
+            } else {
+                CUSTOM_GAS_TOKEN.safeIncreaseAllowance(erc20Gateway, amount);
```
Comment (mrice32) [inline contracts/chain-adapters/Arbitrum_CustomGasToken_Adapter.sol:257]: nit: should this be `requiredL1TokenTotalFeeAmount` instead of `amount`?

**Review by mrice32 — APPROVED** [review]
Looks great! Only a few minor nits

File: contracts/chain-adapters/Arbitrum_CustomGasToken_Adapter.sol (hunk 3 above)
Comment (nicholaspai) [reply contracts/chain-adapters/Arbitrum_CustomGasToken_Adapter.sol:290]: > we could no longer batch these executions with the mainnet spoke pool relayer refund leaf execution

Why not?

File: contracts/chain-adapters/Arbitrum_CustomGasToken_Funder.sol (hunk 4 above)
Comment (nicholaspai) [reply contracts/chain-adapters/Arbitrum_CustomGasToken_Funder.sol:18]: true, will remove

File: contracts/chain-adapters/Arbitrum_CustomGasToken_Adapter.sol (hunk 6 above)
Comment (nicholaspai) [reply contracts/chain-adapters/Arbitrum_CustomGasToken_Adapter.sol]: Yes great catch. I think I'm also missing an increaseAllowance of the l1Token in this else{} branch: 9674ed0f9fbecb6a1fda87e9d5081a91eb47042c

File: contracts/chain-adapters/Arbitrum_CustomGasToken_Adapter.sol (hunk 3 above)
Comment (mrice32) [reply contracts/chain-adapters/Arbitrum_CustomGasToken_Adapter.sol:289]: Sorry, I don't think I said that very clearly. We can batch with this implementation.

If we didn't do the funder contract and just did `transferFrom(msg.sender, ...)`, that would work, but it would break batching since we use the external multicall contract to do batching now, and that contract would be the msg.sender (and contains no tokens).

**Review by mrice32 — APPROVED** [review]
LGTM!
